exercises/sending-signals-external/solution/workflow.py

```diff
@@ -0,0 +1,86 @@
+import asyncio
+from datetime import timedelta
+
+from temporalio import workflow
+from temporalio.exceptions import ApplicationError
+
+# Import activity, passing it through the sandbox without reloading the module
+with workflow.unsafe.imports_passed_through():
+    from activities import PizzaOrderActivities
+    from shared import Bill, OrderConfirmation, PizzaOrder
+
+@workflow.defn
+class PizzaOrderWorkflow:
+    def __init__(self) -> None:
+        self._pending_confirmation: asyncio.Queue[str] = asyncio.Queue()
+
+    @workflow.run
+    async def order_pizza(self, order: PizzaOrder) -> OrderConfirmation:
+        workflow.logger.info(f"order_pizza workflow invoked")
+
+        address = order.address
+
+        total_price = 0
+        for pizza in order.items:
+            total_price += pizza.price
+
+        distance = await workflow.execute_activity_method(
+            PizzaOrderActivities.get_distance,
+            address,
+            start_to_close_timeout=timedelta(seconds=5),
+        )
+
+        if order.is_delivery and distance.kilometers > 25:
+            error_message = "customer lives outside the service area"
+            workflow.logger.error(error_message)
+            raise ApplicationError(error_message)
+
+        workflow.logger.info(f"distance is {distance.kilometers}")
+
+        while True:
+            await workflow.wait_condition(
+                lambda: not self._pending_confirmation.empty()
+            )
+
+            while not self._pending_confirmation.empty():
+                bill = Bill(
+                    customer_id=order.customer.customer_id,
+                    order_number=order.order_number,
+                    description="Pizza order",
+                    amount=total_price,
+                )
+                confirmation = await workflow.execute_activity_method(
+                    PizzaOrderActivities.send_bill,
+                    bill,
+                    start_to_close_timeout=timedelta(seconds=5),
+                )
+                return confirmation
+
+    @workflow.signal
+    async def fulfill_order_signal(self, success: bool) -> None:
+        if success == True:
+            await self._pending_confirmation.put(True)
+
+
+@workflow.defn
+class FulfillOrderWorkflow:
+    @workflow.run
+    async def fulfill_order(self, order: PizzaOrder):
+        workflow.logger.info(f"fulfill_order workflow invoked")
+
+        await workflow.execute_activity_method(
+            PizzaOrderActivities.make_pizzas,
+            order,
+            start_to_close_timeout=timedelta(seconds=5),
+        )
+
+        await workflow.execute_activity_method(
+            PizzaOrderActivities.deliver_pizzas,
+            order,
+            start_to_close_timeout=timedelta(seconds=5),
+        )
+
+        handle = workflow.get_external_workflow_handle("pizza-workflow-order-XD001")
```

Comment:
Arguably a two-workflow solution to this is brittle. A single workflow that does both things would be a better approach IMO. I understand that we're trying to demonstrate external signals here, but this actually shows perfectly why external signals aren't used that commonly because in many cases arbitrary workflow separation may not be the best thing.